typescript/nextjs/src/app/page.tsx

```diff
@@ -0,0 +1,102 @@
+import { Handlers } from "@/substreams/types";
+import { objectToJsonString } from "@/substreams/util";
+import { useQueue } from "@uidotdev/usehooks";
+import { SPKG } from "@/substreams/constants";
+
+export default function Home() {
+  
+  // Store blocks on a queue
+  const { add, queue } = useQueue([]);
+  // Keep state of errors
+  const [error, setError] = useState<string|null>(null);
+  // Keep state of cursor
+  const [cursor, setCursor] = useState<string|null>(null);
+
+  const blockScopeDataHandler = async (response: BlockScopedData, registry: IMessageTypeRegistry) => {
+    const output = response.output?.mapOutput;
+    const cursor = response.cursor;
+
+    if (output !== undefined) {
+        const message = output.unpack(registry);
+        if (message === undefined) {
+            throw new Error(`Failed to unpack output of type ${output.typeUrl}`);
+        }
+
+        // Cursor writing MUST happen after you have successfully processed the message. Otherwise, you risk "skipping" data.
+        const outputAsJson = output.toJson({typeRegistry: registry});
+
+        // Add output to the queue
+        add(JSON.stringify(outputAsJson));
+        
+        await writeCursor(cursor);
+        setCursor(cursor);
+    }
+  }
+
+  /*
+      Handle BlockUndoSignal messages.
+      You will receive this message after a fork has happened in the blockchain.
+
+      Because of the fork, you have probably read incorrect blocks in the "handleBlockScopedDataMessage" function,
+      so you must rewind back to the last valid block.
+  */
+  const blockUndoSignalHandler = async (response: BlockUndoSignal) => {
+    const lastValidBlock = response.lastValidBlock;
+    const lastValidCursor = response.lastValidCursor;
+    
+    /* The blockchain you are streaming from undo 1 or more blocks and you must now handle that case.
+      The field `response.message.<last_valid_block>` contains the last valid block, you must undo whatever
+      has been done prior that (so for data where `block_number > last_valid_block`). Once undo, you must also
+      write the `response.message.<last_valid_cursor>`. In this example, we just print the undo signal and write the cursor.
+    */
+    console.log(`Blockchain undo 1 or more blocks, returning to valid block #${lastValidBlock?.number} (${lastValidBlock?.id})`);
+
+    await writeCursor(lastValidCursor);
+  }
+
+  const progressHandler  = async (message: ModulesProgress) => {
+    console.log(`Progress: ${objectToJsonString(message)}`)
+  }
+
+  const createHandlers = (): Handlers => {
+    return new Handlers(blockScopeDataHandler, blockUndoSignalHandler, progressHandler)
+  }
+
+  useEffect(() => {
+    const executeSubstreams = async () => {
+      try {
+        await startSubstreams(createHandlers());
+      } catch (e) {
+        setError(e);
+        console.log(e)
+      }
+    }
+
+    setCursor(getCursor())
+    executeSubstreams();
+  }, []);
```

Comment:
The `useEffect` has some lifecycle handler when the component unmounts. It should be used to cancel the active "stream". 

https://legacy.reactjs.org/docs/hooks-effect.html#effects-with-cleanup

We should properly terminate the hook when the `useEffect` cleanup is called.